File: test/facade/modelmapper/SearchResultsModelMapperTest.scala
package facade.modelmapper

import org.scalatest.FunSuite
import org.mockito.Mockito._
import org.scalatest.mock.MockitoSugar.mock
import domain.{MockPerson, Person, Team, MockTeam}
import util.CollectionMatcher.contains
import org.hamcrest.MatcherAssert.assertThat
import model.PersonSearchResultMatcher.aPersonSearchResult
import model.TeamSearchResultMatcher.aTeamSearchResult

import scala.collection.JavaConverters._

class SearchResultsModelMapperTest extends FunSuite {
    val searchResultsModelMapper = new SearchResultsModelMapper()

    test("should map persons into person search results") {
        val team = mock[Team]
        when(team.name).thenReturn("a team")

        val fred1 : Person = new MockPerson("1", "fred smith", "pic1.img", team)
        val fred2 : Person = new MockPerson("2", "fred bailey", "pic3.img", team)

        val result = searchResultsModelMapper.map(List(fred1, fred2).asJava, List.empty[Team].asJava)
        assert(result.teams.isEmpty, "should map empty list of teams as empty list")
        assertThat(result.persons, contains(
            aPersonSearchResult.named("fred smith").withPicture("pic1.img").inTeam("a team"),
            aPersonSearchResult.named("fred bailey").withPicture("pic3.img").inTeam("a team")
        ))
    }

    test("should map teams into team search results") {
        val ecom : Team= new MockTeam("1", "Ecom", "blue", List.empty[Person].asJava)
        val anotherTeam : Team = new MockTeam("2", "another team", "red", List.empty[Person].asJava)

        val result = searchResultsModelMapper.map(List.empty[Person].asJava, List(ecom, anotherTeam).asJava)
        assertThat(result.teams, contains(
            aTeamSearchResult.named("Ecom"),
            aTeamSearchResult.named("another team")
        ))
    }

    test("should map person's team name as empty string if person is not in a team") {
        val team = mock[Team]
        when(team.name).thenReturn("")

        val teamless : Person = new MockPerson("1", "teamless", "teamless.img", team)

        val result = searchResultsModelMapper.map(List(teamless).asJava, List.empty[Team].asJava)
        assertThat(result.persons, contains(
            aPersonSearchResult.inTeam("")
        ))
    }
}

File: test/facade/modelmapper/SearchResultsModelMapperTest.java
package facade.modelmapper;

import domain.MockPerson;
import domain.MockTeam;
import domain.Person;
import domain.Team;
import model.DefaultSearchResultsModel;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static model.PersonSearchResultMatcher.aPersonSearchResult;
import static model.TeamSearchResultMatcher.aTeamSearchResult;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static util.CollectionMatcher.contains;

public class SearchResultsModelMapperTest {

    SearchResultsModelMapper searchResultsModelMapper = new SearchResultsModelMapper();

    @Test
    public void should_map_persons_into_person_search_results(){

        Team team = mock(Team.class);
        when(team.name()).thenReturn("a team");

        MockPerson fred1 = new MockPerson("1", "fred smith", "pic1.img", team);
        MockPerson fred2 = new MockPerson("2", "fred bailey", "pic3.img", team);

        List<Person> personList = new ArrayList<>();
        personList.add(fred1);
        personList.add(fred2);
        List<Team> teamList = new ArrayList<>();

        DefaultSearchResultsModel result = searchResultsModelMapper.map(personList, teamList);
        assertThat(result.persons, contains(
                aPersonSearchResult().named("fred smith").withPicture("pic1.img").inTeam("a team"),
                aPersonSearchResult().named("fred bailey").withPicture("pic3.img").inTeam("a team")
        ));
    }

    @Test

    public void should_map_teams_into_team_search_results() {

        List<Person> list1 = new ArrayList<>();
        List<Person> list2 = new ArrayList<>();

        MockTeam ecom = new MockTeam("1", "Ecom", "blue", list1);
        MockTeam anotherTeam = new MockTeam("2", "another team", "red", list2);

        List<Person> list3 = new ArrayList<>();
        List list4 = new ArrayList<>();

        list4.add(ecom);
        list4.add(anotherTeam);

        DefaultSearchResultsModel result = searchResultsModelMapper.map(list3, list4);
        assertThat(result.teams, contains(
                aTeamSearchResult().named("Ecom"),
                aTeamSearchResult().named("another team")
        ));
    }

    @Test
    public void should_map_persons_team_name_as_empty_string_if_person_is_not_in_a_team() {
        Team team = mock(Team.class);
        when(team.name()).thenReturn("");

        MockPerson teamless = new MockPerson("1", "teamless", "teamless.img", team);

        List teamlessList = new ArrayList<>();
        List<Team> emptyTeamList = new ArrayList<>();

        teamlessList.add(teamless);

        DefaultSearchResultsModel result = searchResultsModelMapper.map(teamlessList, emptyTeamList);
        assertThat(result.persons, contains(
                aPersonSearchResult().inTeam("")
        ));

    }
}
